### PR TITLE
compare/unified hover include all points at same coordinate

### LINKED ISSUE #4656: compare/unified hover should include all points at same coordinate
In x-unified and x-compare modes, it should never be possible for there to be two data with the same x-coordinate and only one of them is displayed in a hoverlabel. I know compare mode doesn't currently behave this way but I consider it a bug.

The solution proposed by @alexcjohnson seems good to me:

* Step 1: find hover points the way we do now
* Step 2: loop back over traces not in the set of hover points we already found, and see if any of them have a point with position-axis value exactly matching what we found for the other point(s)

### PR DESCRIPTION
**Pull request by antoinerg**
Fixes #4656 

Codepen: https://codepen.io/antoinerg/pen/gOpKwPg

First commit simply wrap existing logic into a function. Second commit changes the existing behavior.

### CODE REVIEW

File: src/components/fx/hover.js [hunk 1]
```diff
@@ -625,6 +628,46 @@ function _hover(gd, evt, subplot, noHoverEvent) {
 
     hoverData.sort(function(d1, d2) { return d1.distance - d2.distance; });
 
+    // If in compare mode, select every point at position
+    if(hoverData[0].length !== 0 &&
+      ['x', 'y'].indexOf(mode) !== -1 &&
```
Comment (archmoj) [inline src/components/fx/hover.js]: concerning
```js
['x', 'y'].indexOf(mode) !== -1 
```
creating a dynamic array of strings just for index checking might be slow.
Could we create a constant e.g.
```js
var XY = ['x', 'y'];
```
at the upper scope and reuse that here?

File: src/components/fx/hover.js [hunk 2]
```diff
@@ -625,6 +628,46 @@ function _hover(gd, evt, subplot, noHoverEvent) {
...
+      hoverData[0].trace.type !== 'splom' // TODO: add support for splom
+    ) {
+        var hd = hoverData[0];
+        var cd0 = hd.cd[hd.index];
+        var isGrouped = (fullLayout.boxmode === 'group' || fullLayout.violinmode === 'group');
```
Comment (archmoj) [inline src/components/fx/hover.js]: Non-blocking:
Even for `group` which is an array of chars, one may use a const variable.

File: src/components/fx/hover.js [hunk 3]
```diff
@@ -625,6 +628,46 @@ function _hover(gd, evt, subplot, noHoverEvent) {
...
+        if(ax.type === 'category') xVal = ax._categoriesMap[xVal];
+        if(ax.type === 'date') xVal = ax.d2c(xVal);
+        if(cd0 && cd0.t && cd0.t.posLetter === ax._id && isGrouped) {
+            xVal += cd0.t.dPos;
+        }
+
+        var yVal = hd.yVal;
+        ax = hd.ya;
+        if(ax.type === 'category') yVal = ax._categoriesMap[yVal];
+        if(ax.type === 'date') yVal = ax.d2c(yVal);
```
Comment (archmoj) [inline src/components/fx/hover.js]: Here is the second time checking those types.
Let's make 
```js
var isCategory = ax.type === 'category';
var isDate = ax.type === 'date';
```
or use of a `switch` & `case` or put them in one `if` statement.

File: src/components/fx/hover.js (hunk 1 above)
Comment (alexcjohnson) [reply src/components/fx/hover.js]: or even `var XY = {x: 1, y: 1}` -> `if(XY[mode])` 😅

File: src/components/fx/hover.js [hunk 4]
```diff
@@ -625,6 +628,46 @@ function _hover(gd, evt, subplot, noHoverEvent) {
...
+        if(cd0 && cd0.t && cd0.t.posLetter === ax._id && isGrouped) {
+            yVal += cd0.t.dPos;
+        }
+
+        findHoverPoints(xVal, yVal);
+
+        // Remove duplicated hoverData points
+        // note that d3 also filters identical points in the rendering steps
+        // TODO: use ES6 map
```
Comment (alexcjohnson) [inline src/components/fx/hover.js]: I don't think we can make that a TODO just yet - perhaps if & when we convert the whole syntax to ES6 we can make sure we include polyfills for this and other feature, THEN we can use it whenever we wish...

File: src/components/fx/hover.js (hunk 3 above)
Comment (antoinerg) [reply src/components/fx/hover.js]: It's not the same axis object the second time around!

**Comment by antoinerg** [comment]
> This looks great. Just a few nonblocking comments.

Great! I'll merge as soon as the tests pass

File: src/components/fx/hover.js (hunk 1 above)
Comment (antoinerg) [reply src/components/fx/hover.js]: Done in https://github.com/plotly/plotly.js/pull/4664/commits/b8b9395f573a8b70e8263aac2ef154de49d6de5f

File: src/components/fx/hover.js (hunk 3 above)
Comment (archmoj) [reply src/components/fx/hover.js]: Oops. Resolving.

**Comment by archmoj** [comment]
Nicely done! 
:dancer: :dancer:

**Pull request merged**